### ISSUE TITLE
Using service containers

### ISSUE THREAD

**Issue by jenstroeger** [step-security/harden-runner#152]
I tried out the `harden-runner` action (based on [this repo](https://github.com/jenstroeger/python-package-template/blob/ba77a96deec32e351ecd931e2c15843522be36ad/.github/workflows/build.yaml#L22-L25)) and
```yaml
with:
  egress-policy: audit
```
and it worked, until I added a [PostgreSQL service container](https://docs.github.com/en/actions/using-containerized-services/creating-postgresql-service-containers) to run a few tests. It looks like traffic to that container is blocked? I tried to add
```yaml
allowed-endpoints: >
  localhost:5432  # `postgres:5432` doesn’t work either
```
but neither of these two worked. It’s a private organization and I don’t have the privileges to install the [app](https://github.com/marketplace/harden-runner-app) to check the egress audit log.

I disabled the step, and all tests pass just fine. How do you recommend to proceed?

Much thanks!

**Comment by Fich0Gl** [comment]
other variants of this are 127.0.0.1:5432 and 0.0.0.0:5432. I don't know if it is worth trying, the best way to debug this would be to add the Harden Runner App to check whether the port is open or not

**Comment by varunsh-coder** [comment]
@h0x0er can you please try to repro this issue on a public repo? You can use the workflow from here: https://docs.github.com/en/actions/using-containerized-services/creating-postgresql-service-containers#running-jobs-directly-on-the-runner-machine

**Comment by jenstroeger** [comment]
Thank you, @h0x0er.

One question: why would the PostgreSQL container stop running, thus warranting the automatic restart? I wasn’t able to find details on the health options but perhaps they’re insufficient if subsequent jobs take too long? I mean `5` retries on a `5s` timeout is 25 seconds, not sure where exactly the interval comes in here 🤔

**Comment by varunsh-coder** [comment]
> Thank you, @h0x0er.
> 
> One question: why would the PostgreSQL container stop running, thus warranting the automatic restart? I wasn’t able to find details on the health options but perhaps they’re insufficient if subsequent jobs take too long? I mean `5` retries on a `5s` timeout is 25 seconds, not sure where exactly the interval comes in here 🤔

@jenstroeger I can answer your question. 

`harden-runner` Github Action installs an agent to monitor the build process. That agent runs a DNS proxy and additional monitoring on the Ubuntu VM. As part of that, the agent needs to restart the docker daemon. You can see the code below:

https://github.com/step-security/agent/blob/main/dnsconfig.go#L169

Normally, by the time all this happens, in the `pre harden-runner` step, no images have started to run and no workflow steps have run. But in this scenario, the PostgreSQL container is started before the `pre harden-runner` step. So when the docker daemon is restarted, this image stops running and doesn't restart on its own. 

I hope this answers your question. 

@h0x0er is trying to figure out if as part of restarting docker daemon, we can restart all images that were already running. But if we cannot figure that out, we will need to add documentation to add the `--restart always` argument in this scenario.

**Comment by varunsh-coder** [comment]
@h0x0er was able to figure out a way to restart existing running containers as part of the docker daemon restart. I will test out the changes and release next week. After new version is released, you will not need to add `--restart always`. It should just work as expected.

**Comment by jenstroeger** [comment]
@varunsh-coder thanks for the update! I’ll wait for the next release and then update on my end, and I’ll let you know whether it works.

**Comment by varunsh-coder** [comment]
This is fixed in the latest release `v1.4.5` with tag `dd2c410b088af7c0dc8046f3ac9a8f4148492a95`. 
You should not need any workaround for this to work. 
We have also added an integration test for it. Here is an example workflow run:https://github.com/harden-runner-canary/postgres-testing/runs/7810960312?check_suite_focus=true#step:9:10 and insights URL: https://app.stepsecurity.io/github/harden-runner-canary/postgres-testing/actions/runs/2848365031

**Comment by jenstroeger** [comment]
> You should not need any workaround for this to work.

Confirming that the change works.
